src/rqt_service_caller/service_caller_widget.py

```diff
@@ -262,23 +308,37 @@ def _evaluate_expression(self, expression, slot_type):
     def on_call_service_button_clicked(self):
         self.response_tree_widget.clear()
 
-        request = self._service_info['service_class']._request_class()
-        self.fill_message_slots(request, self._service_info['service_name'],
-                                self._service_info['expressions'], self._service_info['counter'])
-        try:
-            response = self._service_info['service_proxy'](request)
-        except rospy.ServiceException as e:
+        request = self._service_info['service_class'].Request()
+        self.fill_message_slots(
+            request, self._service_info['service_name'], self._service_info['expressions'],
+            self._service_info['counter'])
+
+        cli = self._node.create_client(
+            self._service_info['service_class'],  self._service_info['service_name'])
+
+        while not cli.wait_for_service(timeout_sec=3.0):
+            qWarning(
+                'ServiceCaller.on_call_service_button_clicked()'
+                'Service ({}, {}) not available'.format(
+                    self._service_info['service_name'],
+                    self._service_info['service_class']))
+
+        future = cli.call_async(request)
+        rclpy.spin_until_future_complete(self._node, future)
```

Comment:
So currently self._node is not spinning when it is created. In order to get subscribers to work, I need to make a PR that spins the node in a separate QThread. How will this logic change?

Reply:
Good point. A few of the plugins may need to change once we have the executor code up and running. I spent some time architecting what this might look like last night. I stopped when I saw that you already have a branch in progress. I'm not sure the state of your branch but it seems to me that the python should model itself after the cpp code. Ie. the plugin provider spins off a separate thread with an executor which manages callbacks for the single context node. The complexity here is that the Qt code already has a spin of it's own in qt_gui main.py so we would have to ensure that the threads both exit cleanly on shutdown.

So far as the rqt_service_caller is concerned, the correct implementation depends slightly on how you set up the executor.